src/Components/Filters.js

```diff
@@ -11,44 +12,133 @@ const Button = props => {
-
-  const onLevelChange = pressedLevel => {
-    const newLevel = pressedLevel === level ? undefined : pressedLevel;
-    setLevel(newLevel);
+class Filters extends React.Component {
+  state = {
+    selectedLevel: undefined,
+    selectedGoal: undefined,
+    numberOfPlayers: undefined,
+    displayedDrills: this.props.route.params.initialData,
   };
 
-  useEffect(() => {
-    const newData = level ? props.initialData.filter(drill => drill.level === level) : props.initialData;
-    props.onFiltered(newData);
-  }, [level]);
+  onPressedChange(target, value) {
+    this.setState(prevState => {
+      const newValue = value === prevState[target] ? undefined : value;
+      return {
+        [target]: newValue,
+      };
+    }, this.applyFilters);
+  }
 
-  return (
-    <View style={styles.filters}>
-      <Text style={styles.filterTitle}>Level</Text>
-      <View style={styles.filter}>
-        <Button title="Beginner" onPress={() => onLevelChange('beginner')} active={level === 'beginner'} />
-        <Button title="Intermediate" onPress={() => onLevelChange('intermediate')} active={level === 'intermediate'} />
-        <Button title="Advanced" onPress={() => onLevelChange('advanced')} active={level === 'advanced'} />
+  onInputChange(target, value) {
+    this.setState({ [target]: value }, this.applyFilters);
+  }
+
+  applyFilters() {
+    const { selectedLevel, selectedGoal, numberOfPlayers } = this.state;
+    let newData = this.props.route.params.initialData;
+
+    if (selectedLevel) newData = newData.filter(drill => drill.level === selectedLevel);
+    if (selectedGoal) newData = newData.filter(drill => drill.goals.includes(selectedGoal));
+    if (numberOfPlayers) newData = newData.filter(drill => drill.minimalPlayersNumber <= numberOfPlayers);
+
+    this.setState({ displayedDrills: newData });
+  }
+
+  componentDidMount() {
+    this.props.route.params.onFiltered(this.state.displayedDrills);
+    this.props.navigation.setOptions({
+      headerRight: () => (
+        <TouchableOpacity
+          style={styles.headerButton}
+          onPress={() => {
+            this.props.route.params.onFiltered(this.state.displayedDrills);
+            this.props.navigation.goBack();
+          }}
+        >
+          <Text style={styles.headerButtonText}>✓</Text>
+        </TouchableOpacity>
+      ),
+    });
+  }
+
+  render() {
+    const { selectedLevel, selectedGoal, numberOfPlayers } = this.state;
+    return (
+      <View style={styles.wrapper}>
+        <Text testID="availableDrills" style={styles.counter}>
+          {this.state.displayedDrills.length} drills available
+        </Text>
+        <View style={styles.filters}>
+          <Text style={styles.filterTitle}>Level</Text>
+          <View style={styles.filter}>
+            {Object.values(Levels).map(level => (
+              <Button
+                key={level}
+                title={level}
+                onPress={this.onPressedChange.bind(this, 'selectedLevel', level)}
```

Comment:
It's a bit annoying because I would have to bind quite a lot of functions in the constructor

... which means the performance issues would be even more annoying if we keep redefining functions here, I guess?

Reply:
Yeah, the other solution is to make an arrow function.

Reply:
It's done
Tell me if it's ok :)